src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHelpers.cs

```diff
@@ -232,33 +200,67 @@ internal static TarEntryType GetCorrectTypeFlagForFormat(TarEntryFormat format,
 
         // Receives a byte array that represents an ASCII string containing a number in octal base.
         // Converts the array to an octal base number, then transforms it to ten base and returns it.
-        internal static int GetTenBaseNumberFromOctalAsciiChars(Span<byte> buffer)
+        internal static int ParseOctalAsInt32(ReadOnlySpan<byte> buffer)
         {
-            string str = GetTrimmedAsciiString(buffer);
-            return string.IsNullOrEmpty(str) ? 0 : Convert.ToInt32(str, fromBase: 8);
+            buffer = TrimEndingNullsAndSpaces(buffer);
+
+            uint value = 0;
+            foreach (byte b in buffer)
+            {
+                uint digit = (uint)(b - '0');
+                if (digit >= 8)
+                {
+                    ThrowInvalidNumber();
+                }
+
+                value = checked((value * 8u) + digit);
+            }
+
+            return (int)value;
         }
 
         // Receives a byte array that represents an ASCII string containing a number in octal base.
         // Converts the array to an octal base number, then transforms it to ten base and returns it.
-        internal static long GetTenBaseLongFromOctalAsciiChars(Span<byte> buffer)
+        internal static long ParseOctalAsInt64(ReadOnlySpan<byte> buffer)
```

Comment:
does generic math allow combining with ParseOctalAsInt32 ? with an out parameter of T?

Reply:
That should be doable...